controller/array_action/array_mediator_svc.py

```diff
@@ -443,29 +479,45 @@ def _stop_and_delete_fcmap(self, fcmap_id):
         self._stop_fcmap(fcmap_id)
         self._delete_fcmap(fcmap_id, force=True)
 
-    def _delete_snapshot(self, target_cli_volume):
-        self._stop_and_delete_fcmap(target_cli_volume.FC_id)
-        self._delete_volume_by_name(target_cli_volume.name)
+    def _delete_all_fcmaps_as_source_if_exist(self, snapshot_name):
+        snapshot_as_source_fcmaps = self._get_fcmaps(snapshot_name, ENDPOINT_TYPE_SOURCE)
+        unfinished_fcmaps = [fcmap.name for fcmap in snapshot_as_source_fcmaps
+                             if fcmap.status != FCMAP_STATUS_DONE]
+        if unfinished_fcmaps:
+            raise controller_errors.SnapshotIsStillInUseError(snapshot_name, unfinished_fcmaps)
+        for fcmap in snapshot_as_source_fcmaps:
+            self._delete_fcmap(fcmap.id, force=False)
+
+    def _delete_snapshot(self, cli_volume):
+        fcmap_id = cli_volume.FC_id
+        snapshot_name = cli_volume.name
+        if fcmap_id == 'many':
+            self._delete_all_fcmaps_as_source_if_exist(snapshot_name)
+        fcmap = self._get_fcmap_as_target_if_exists(snapshot_name)
+        if not fcmap:
+            raise controller_errors.SnapshotNotFoundError(snapshot_name)
```

Comment:
is that what it means if "not fcmap"? that no Snapshot found?

Reply:
it's not a snapshot if it doesn't have a mapping to it, and this is the existing exception in the mediator interface.
do you think we should add "snapshot is volume" error/s for `delete_snapshot` in the interface?

Reply:
well, I dont understand how do we really get to here if its not a snapshot..
the mediator server already verified that, right?

Reply:
not that I'm aware of, but let me know if I missed it

Reply:
no verification.. but its odd, it will raise on snapshot object.. the k8s snapshot will say there is no snapshot..

Reply:
I agree that this exception is not ideal for this scenario.
so what do you suggest?

1. adding an exception in the interface?
2. delete it without an error?
3. leave it as it is?
4. something else?

Reply:
btw, what if its 'many' there? isn't a bad case?

Reply:
I think exception and deletion

Reply:
yes, `many` should cause a storage exception on deletion attempt

Reply:
do you mean that we should delete and then raise an exception?
if so, the second retry would succeed, so the exception would not mean much

Reply:
it will if we will do it before checking on target (if finished copying), if it not a snapshot, I cant see why to check it..

Reply:
in other hand, if it not a snapshot, maybe its better if we wont delete it..

Reply:
do you think it's mandatory for GA?
if not, then we would have more time to think about it until CSI-1556

Reply:
no, really not mandatory

Reply:
first verify its a snapshot, later delete all flashcopies

Reply:
and if it's not a snapshot, should it still be deleted?

Reply:
and is this current error ok?

Reply:
if not snapshot, I think its not ok to remove flashcopies.. and the error seems right

Reply:
so if there are any action items here, please specify.
if not, please resolve.

Reply:
first check if its a snapshot, if it is, remove flashcopies and continue, else, return snapshot not found and dont touch. that is my opinion.

Reply:
so what exactly is missing in the current code?

Reply:
self._delete_all_fcmaps_as_source_if_exist(snapshot_name) after if not fcmap:

Reply:
but we already have this call right before `if not fcmap:`.
could you clarify what you mean?